content/gocourse-2019-05-melbourne.slide

```diff
@@ -813,49 +816,106 @@ Embedding types introduces the possiblity for naming conflicts. Go follows some
 .caption J. D. Frazer "Illiad", User Friendly - [[https://en.wikipedia.org/wiki/User_Friendly#Crud_Puppy][Crud Puppy]]
 
 
-* Errors (1/3)
+* Errors in Go
 
-"Errors are values."  - [[https://go-proverbs.github.io/][Rob Pike]]
-
-The `error` type is a built-in interface:
+- `error` a builtin interface ([[https://golang.org/src/builtin/builtin.go?s=11246:11286#L251][source: error]])
+- Check errors explicitly
+- The [[https://golang.org/pkg/errors/][`errors`]] package has functions for creating new error type
+- Creation of new error type can be done in a few ways:
 
-	type error interface {
-	    Error() string
+	func Sqrt(f float64) (float64, error) {
+		if f < 0 {
+			return 0, errors.New("math: square root of negative number")
+		}
+		// implementation
 	}
+- Or use formatted error message using `fmt.Errorf`:
+	if f < 0 {
+		return 0, fmt.Errorf("math: square root of negative number %g", f)
+	}
+
+* Errors in Go (continued)
 
-By convention, functions often return `error` as the second/last return value.
-The calling code should handle errors by testing whether the error equals nil:
+- Custom error types can be created. Example from `net` package [[https://golang.org/pkg/net/#AddrError][net/AddrError]]:
+	type AddrError struct {
+		Err  string
+		Addr string
+	}
 
-	func getCat(strID string) (*Cat, error) {
-		id, err := strconv.Atoi(strID)
-		if err != nil {
-		    return nil, CatError(err)
+	func (e *AddrError) Error() string {
+		if e == nil {
+			return "<nil>"
+		}
+		s := e.Err
+		if e.Addr != "" {
+			s = "address " + e.Addr + ": " + s
 		}
-		return catStore.cats[id], nil
+		return s
 	}
+- The error values can be logged and printed:
 
+	_, err := Sqrt(-1); err != nil {
+		log.Println(err)
+	}
 
-* Errors (2/3)
 
-"Don't panic." - [[https://go-proverbs.github.io/][Rob Pike]]
+* Error Handling
 
-Panic should be used when something went unexpectedly wrong.
-For instance, initialise a regular expression with [[https://golang.org/pkg/regexp/#MustCompile][regexp.MustCompile]] if the pattern is a string constant.
-`MustCompile` will `panic` at initialisation.
+- Check for the presence of an error (compare with nil)
+- Do not compare error variables ([[https://goplay.space/#mgw9QAmeLUP][samples/errors4.go]])
+- Errors are values so it can be programmed ([[https://goplay.space/#kbBftimqvuF][samples/errors5.go]])
+- Handle errors gracefully ([[https://goplay.space/#GkfuGGvpL0F][A useful pattern (samples/errors6.go)]])
+- To check if an `error` is of certain type use custom error types with type assertion
```

Comment:
To check if an `error` is of _a_ certain type use custom error types with type assertion

This is just one of a number of ways. You could implement an error with a Code/Kind field: e.g. [error handling in upspin](https://commandcenter.blogspot.com/2017/12/error-handling-in-upspin.html)

Great read:
https://go.googlesource.com/proposal/+/master/design/go2draft-error-values-overview.md